### PR TITLE
Address misaligned match lines on Simon front and left pieces

### PR DESCRIPTION
**Pull request by nicholasdower**
# Background

I recently attempted the Simon pattern and ran into a small issue with pattern matching. As I am very new to sewing and this was my first time ever sewing from a pattern, I think it is likely that I have misunderstood something or made a mistake.  Would you mind reviewing my issue and proposed change to see whether it makes sense?

# Problem Description

The issue is with the pattern matching between the two front pieces. When my shirt is fully buttoned, the pattern alignment is off by around six millimeters:
image

You can see how the pattern should match if I unbutton the shirt and shift the sides:
image

Looking at the generated pattern pieces and aligning the button and buttonhole lines, I see that the match lines are indeed misaligned:
image

As far as I can tell, this is caused by the fact that the position of the match line on the front right piece is based solely on the width of the button placket. I believe it should be based on the size of the buttonhole placket.

# Approach 

In this change I have attempted to add a separate match line to the front right piece. It is based on the buttonhole placket width and is usually in a different position that the line marking the edge of the button placket. I have also updated both front pieces to only display match lines if the placket on the other side is not separate. As far as I can tell, match lines are never displayed on separate plackets.

# Testing

I was a little unsure which combinations of plackets I needed to support. I believe there are 16 theoretically possible placket combinations: separate/attached buttonhole * classic/seamless buttonhole * separate/attached button * classic/seamless button. In practice it seems all combinations involving separate plackets can be ignored since they do not include match lines. That leaves the following combinations:

## attached-classic buttonhole, attached-classic-button

image

## attached-classic buttonhole, attached-seamless-button

image

## attached-seamless buttonhole, attached-classic-button

image

## attached-seamless buttonhole, attached-seamless-button
image

## Close lines

I also tested the behavior when the match line and placket edge are very close:
image
image

# Open Questions

I wasn't sure what to do about notches. I see that notches are normally displayed on the match lines and on the edges of the plackets, but as far as I can tell these notches do not match with any other notches. I have tentatively removed the notches I could not understand from the button placket. I did not remove the corresponding notches from the buttonhole placket.. I could use some guidance here.

~~I didn't find any way to have the match line meet the curve of the neck. Please let me know if there is a way to improve this.~~ (Found `utils.lineIntersectsCurve`. Added a separate commit for this.)

### CODE REVIEW

**Comment by joostdecock** [comment]
I'll need to have a look at this in detail (on my phone right now) but I can already say I'm very impressed with the quality of this PR 👌

**Comment by joostdecock** [comment]
I am not entirely certain why I ended up putting that match line there, but I was wondering whether we'd not make life a lot easier on ourselves if we simply moved the *match fabric here* text to center front (in other words, the line on which the buttons and buttonholes lie.

Because if you match fabric on the CF line, it will always match regardless of how wide the plackets are.

I feel that might save us some headaches in the future as there's already a lot of different pocket combinations and this would remove one thing that needs to be checked for each combo.

I'd like to hear your thoughts on this @nicholasdower as I don't want to be dismissive of your PR. I'm just wondering if we should use this opportunity to simplify things.

**Comment by nicholasdower** [comment]
> I am not entirely certain why I ended up putting that match line there, but I was wondering whether we'd not make life a lot easier on ourselves if we simply moved the _match fabric here_ text to center front (in other words, the line on which the buttons and buttonholes lie.
> 
> Because if you match fabric on the CF line, it will always match regardless of how wide the plackets are.
> 
> I feel that might save us some headaches in the future as there's already a lot of different pocket combinations and this would remove one thing that needs to be checked for each combo.
> 
> I'd like to hear your thoughts on this @nicholasdower as I don't want to be dismissive of your PR. I'm just wondering if we should use this opportunity to simplify things.

This kind of blew my mind. It makes perfect sense!

Let me make the change real quick. Also, no hard feelings at all. I learned a ton poking around and I think it was worth the effort.

**Comment by nicholasdower** [comment]
Thanks @joostdecock for the quick and thoughtful review. I have made the suggested change. Please take another look.

File: packages/simon/src/frontleft-classic-cuton.js [hunk 1]
```diff
@@ -32,6 +32,12 @@ export default (part) => {
       .move(points.placketCfNeck)
       .line(points.placketCfHem)
       .attr('class', 'help')
+    if (!options.seperateButtonPlacket) {
```
Comment (nicholasdower) [inline packages/simon/src/frontleft-classic-cuton.js:35]: I'm not sure it makes sense to have this check. I am trying to avoid displaying a match line when there is a separate placket but it looks like this setting cannot be relied on for that. If I choose separate placket for either side of the shirt, but also set the corresponding placket style to seamless, the pattern does not actually contain a separate placket. Assuming this is a bug, maybe it makes more sense for me to address that issue first, then add this check. wdyt?

File: packages/simon/src/frontright-classic-cuton.js [hunk 2]
```diff
@@ -25,24 +25,24 @@ export default (part) => {
...
       .line(points.placketBottomOut)
       .attr('class', 'dotted')
     macro('sprinkle', {
       snippet: 'notch',
       on: [
-        'placketTopIn',
-        'placketTopOut',
         'cfNeck',
-        'placketBottomIn',
-        'placketBottomOut',
```
Comment (nicholasdower) [inline packages/simon/src/frontright-classic-cuton.js:45]: Does it make sense to remove these? Maybe these notches were being used for something other than matching the sides? If it does make sense, should I also remove the notches on the buttonhole placket?

File: packages/simon/src/frontleft-classic-cuton.js (hunk 1 above)
Comment (nicholasdower) [reply packages/simon/src/frontleft-classic-cuton.js:35]: The idea would be to prevent the user from choosing seamless when a separate placket has been selected.

Comment (joostdecock) [reply packages/simon/src/frontleft-classic-cuton.js:35]: Yeah so the issue here is that *seamless* is a style without any seam line at the placket, and thus it is not possible to have a separate placket and seamless style. And in practice, the seamless setting takes precedent.

What's needed here is some way to designate options dependent on others and hide them in the frontend when they become irrelevant. So that when the user picks a 'seamless' placket style, the option to make it separate is no longer available.

That is not something we currently support, but it's on our roadmap now, and given that we're also building a new frontend, this is :100: something I want to add.

As to the matter at hand, we can leave the check in.

Comment (joostdecock) [reply packages/simon/src/frontleft-classic-cuton.js:35]: Forgot the link to the roadmap: #1278

File: packages/simon/src/frontright-classic-cuton.js (hunk 2 above)
Comment (joostdecock) [reply packages/simon/src/frontright-classic-cuton.js:45]: These notches are here to help people fold the placket during construction.

Comment (nicholasdower) [reply packages/simon/src/frontright-classic-cuton.js:45]: Ok, I will revert these changes. I had the impression that notches were only used to match two pieces of fabric. That said, I have literally never seen a notch before this pattern.

Comment (nicholasdower) [reply packages/simon/src/frontright-classic-cuton.js]: Added a commit reverting the notch removal.

**Comment by joostdecock** [comment]
LGTM!

**Comment by joostdecock** [comment]
@all-contributors please add @nicholasdower for code

**Comment by joostdecock** [comment]
Thank you very much for this PR @nicholasdower it was an absolute pleasure to receive.

Also, this qualifies for our bug bounty project, so if you get me your postal address, I'd be happy to send you some stickers. You can either DM me on https://discord.freesewing.org/ or email me: `joost@joost.at`

Thanks again for your contribution, really great work :pray:

**Comment by nicholasdower** [comment]
Thanks @joostdecock for the thoughtful review!

**Comment by joostdecock** [comment]
Fix published in 2.19.9: https://www.npmjs.com/package/@freesewing/simon/v/2.19.9
Being deployed on freesewing.org as we speak: https://app.netlify.com/sites/fs-org-en/deploys/61db28e0d8f7420008225c57

**Comment by nicholasdower** [comment]
> Fix published in 2.19.9: https://www.npmjs.com/package/@freesewing/simon/v/2.19.9 Being deployed on freesewing.org as we speak: https://app.netlify.com/sites/fs-org-en/deploys/61db28e0d8f7420008225c57

So rewarding to have a fix deployed same day. Thank you for taking the time.
